File: 300_longest_increasing_subsequence/longest_increasing_subsequence.py
from typing import List
from bisect import bisect_left

class Solution:
    def longest_increasing_subsequence(self, nums: List[int]) -> int:
        sub = []
        for x in nums:
            i = bisect_left(sub, x)
            if i == len(sub):
                sub.append(x)
            else:
                sub[i] = x
        return len(sub)


if __name__ == "__main__":
    nums = [10, 9, 2, 5, 3, 7, 101, 18]
    solution = Solution()
    print(solution.longest_increasing_subsequence(nums))

File: 300_longest_increasing_subsequence/longest_increasing_subsequence.go
package main

import (
	"fmt"
	"sort"
)

type Solution struct{}

func (s *Solution) longestIncreasingSubsequence(nums []int) int {
	sub := []int{}

	for _, x := range nums {
		i := sort.SearchInts(sub, x)
		if i == len(sub) {
			sub = append(sub, x)
		} else {
			sub[i] = x
		}
	}

	return len(sub)
}

func main() {
	nums := []int{10, 9, 2, 5, 3, 7, 101, 18}
	s := Solution{}
	fmt.Println(s.longestIncreasingSubsequence(nums))
}
